compiler/rustc_trait_selection/src/traits/util.rs

```diff
@@ -507,19 +510,69 @@ impl<'tcx> TypeFolder<TyCtxt<'tcx>> for PlaceholderReplacer<'_, 'tcx> {
 }
 
 pub fn sizedness_fast_path<'tcx>(tcx: TyCtxt<'tcx>, predicate: ty::Predicate<'tcx>) -> bool {
-    // Proving `Sized` very often on "obviously sized" types like `&T`, accounts for about 60%
-    // percentage of the predicates we have to prove. No need to canonicalize and all that for
-    // such cases.
+    // Proving `Sized`/`MetaSized`/`PointeeSized`, very often on "obviously sized" types like
+    // `&T`, accounts for about 60% percentage of the predicates we have to prove. No need to
+    // canonicalize and all that for such cases.
     if let ty::PredicateKind::Clause(ty::ClauseKind::Trait(trait_ref)) =
         predicate.kind().skip_binder()
     {
-        if tcx.is_lang_item(trait_ref.def_id(), LangItem::Sized)
-            && trait_ref.self_ty().is_trivially_sized(tcx)
-        {
+        let sizedness = match tcx.as_lang_item(trait_ref.def_id()) {
+            Some(LangItem::Sized) => SizedTraitKind::Sized,
+            Some(LangItem::MetaSized) => SizedTraitKind::MetaSized,
+            Some(LangItem::PointeeSized) => SizedTraitKind::PointeeSized,
+            _ => return false,
+        };
+
+        if trait_ref.self_ty().has_trivial_sizedness(tcx, sizedness) {
             debug!("fast path -- trivial sizedness");
             return true;
         }
     }
 
     false
 }
+
+/// To improve performance, sizedness traits are not elaborated and so special-casing is required
+/// in the trait solver to find a `Sized` candidate for a `MetaSized` obligation. Returns the
+/// predicate to used in the candidate for such a `obligation`, given a `candidate`.
+pub(crate) fn unelaborated_sizedness_candidate<'tcx>(
+    infcx: &InferCtxt<'tcx>,
+    obligation: &PolyTraitObligation<'tcx>,
+    candidate: PolyTraitPredicate<'tcx>,
+) -> PolyTraitPredicate<'tcx> {
+    use crate::infer::InferCtxtExt;
+    if !infcx.tcx.is_lang_item(obligation.predicate.def_id(), LangItem::MetaSized)
+        || !infcx.tcx.is_lang_item(candidate.def_id(), LangItem::Sized)
+    {
+        return candidate;
+    }
+
+    let drcx = DeepRejectCtxt::relate_rigid_rigid(infcx.tcx);
+    let expected_self_ty = obligation.self_ty();
+    let found_self_ty = candidate.self_ty();
+    let matches = obligation.predicate.polarity() == candidate.polarity()
+        && drcx.args_may_unify(
+            obligation.predicate.skip_binder().trait_ref.args,
+            candidate.skip_binder().trait_ref.args,
+        )
+        && expected_self_ty.bound_vars() == found_self_ty.bound_vars()
+        && {
+            let expected_self_ty =
+                infcx.tcx.instantiate_bound_regions_with_erased(expected_self_ty);
+            let found_self_ty = infcx.tcx.instantiate_bound_regions_with_erased(found_self_ty);
+            infcx.can_eq(obligation.param_env, expected_self_ty, found_self_ty)
+        };
```

Comment:
```suggestion
        && drcx.args_may_unify(
            obligation.predicate.skip_binder().trait_ref.args,
            candidate.skip_binder().trait_ref.args,
        );
```
the `args_may_unify` should be enough here :thinking:

style: please change this to use early returns instead

Reply:
Fixed